env0/resource_module.go

```diff
@@ -0,0 +1,199 @@
+		UpdateContext: resourceModuleUpdate,
+		DeleteContext: resourceModuleDelete,
+
+		Importer: &schema.ResourceImporter{StateContext: resourceModuleImport},
+
+		Schema: map[string]*schema.Schema{
+			"module_name": {
+				Type:             schema.TypeString,
+				Description:      "name of the module",
+				Required:         true,
+				ValidateDiagFunc: NewRegexValidator(`^[0-9A-Za-z](?:[0-9A-Za-z-_]{0,62}[0-9A-Za-z])?$`),
+			},
+			"module_provider": {
+				Type:             schema.TypeString,
+				Description:      "the provider name in the module source",
+				Optional:         true,
+				ValidateDiagFunc: NewRegexValidator(`^[0-9a-z]{0,64}$`),
+			},
+			"repository": {
+				Type:        schema.TypeString,
+				Description: "the repository containing the module files",
+				Required:    true,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Description: "description of the module",
+				Optional:    true,
+			},
+			"token_id": {
+				Type:         schema.TypeString,
+				Description:  "the token id used for integration with GitLab",
+				Optional:     true,
+				ExactlyOneOf: vcsExcatlyOneOf,
+			},
+			"token_name": {
+				Type:         schema.TypeString,
+				Description:  "the token name used for integration with GitLab",
+				Optional:     true,
+				RequiredWith: []string{"token_id"},
+			},
+			"github_installation_id": {
+				Type:         schema.TypeInt,
+				Description:  "The env0 application installation id on the relevant Github repository",
+				Optional:     true,
+				ExactlyOneOf: vcsExcatlyOneOf,
+			},
+			"bitbucket_client_key": {
+				Type:         schema.TypeString,
+				Description:  "The client key used for integration with Bitbucket",
+				Optional:     true,
+				ExactlyOneOf: vcsExcatlyOneOf,
+			},
+			"ssh_keys": {
+				Type:        schema.TypeList,
+				Description: "an array of references to 'data_ssh_key' to use when accessing git over ssh",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type:        schema.TypeMap,
+					Description: "a map of env0_ssh_key.id and env0_ssh_key.name for each module",
+				},
+			},
+		},
+	}
+}
+
+func resourceModuleCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	var payload client.ModuleCreatePayload
+	if err := readResourceData(&payload, d); err != nil {
+		diag.Errorf("schema resource data deserialization failed: %v", err)
+	}
+
+	if len(payload.TokenId) > 0 {
+		payload.IsGitlab = boolPtr(true)
+	}
+
+	module, err := apiClient.ModuleCreate(payload)
+	if err != nil {
+		return diag.Errorf("could not create module: %v", err)
+	}
+
+	d.SetId(module.Id)
+
+	return nil
+}
+
+func resourceModuleRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	module, err := apiClient.Module(d.Id())
+	if err != nil {
+		return diag.Errorf("could not get module: %v", err)
+	}
```

Comment:
We should be handling drifts as well for this case. 
In case error status code is `404` we should call `d.SetId("")` which would basically remove the resource from the TF state and will force TF to try and recreate it (aka reconcile the drift)

For every other status code we should be throwing an error as you did.

Reply:
https://github.com/env0/terraform-provider-env0/pull/279/commits/a59b4bbbb4c4f0f861546b4848b2c55e5040adbb

Fixed also for notification